src/main/java/com/google/solutions/ml/api/vision/AnnotateImagesSimulatorDoFn.java

```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2020 Google LLC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.google.solutions.ml.api.vision;
+
+import com.google.cloud.vision.v1.AnnotateImageResponse;
+import com.google.cloud.vision.v1.AnnotateImageResponse.Builder;
+import com.google.cloud.vision.v1.Feature;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import org.apache.beam.sdk.transforms.DoFn;
+import org.apache.beam.sdk.values.KV;
+
+/**
+ * Image annotation simulation class to test batching logic without incurring Vision API costs.
+ *
+ * It simulates the delay of calling the API and produces a single annotation.
+ */
+public class AnnotateImagesSimulatorDoFn extends
+    DoFn<Iterable<String>, KV<String, AnnotateImageResponse>> {
+  private static final long serialVersionUID = 1L;
+
+  /**
+   * Not used. But the simulation logic can be enhanced if needed to produce annotations
+   * based on the requested features.
+   */
+  private final List<Feature> featureList = new ArrayList<>();
+
+  public AnnotateImagesSimulatorDoFn(List<Feature.Type> featureTypes) {
+    featureTypes.forEach(
+        type -> featureList.add(Feature.newBuilder().setType(type).build()));
+  }
+
+  @ProcessElement
+  public void processElement(@Element Iterable<String> imageUris,
+      OutputReceiver<KV<String, AnnotateImageResponse>> out) {
+    VisionAnalyticsPipeline.numberOfRequests.inc();
+    try {
+      Thread.sleep(500 + (new Random().nextInt(1000)));
```

Comment:
It creates a pattern similar to using the actual APIs with 16 requests per batch. It can be made dynamic/parameterized, but I didn't want to complicate it, esp. because this is a simulation code and not the main branch.

Reply:
Ok makes sense. I'd just add a comment in the code to explain, so the reader understands the reasoning.

Reply:
Added comments.